### DISCUSSION TITLE
Integration in Home Assistant

### DISCUSSION THREAD

**Discussion by terafilgit** [cmcgerty/Sofar2mqtt#26] [Q&A] [answered]
Hello everyone,
I've successfully connected my HYD6000 to the mqtt server, but I don't know how to read data in Home Assistant (HA from now then)...
In HA as MQTT settings I have putted: 
- Broker: core-mosquitto
- Port: 1883
- Username: homeassistant
- Password: the one suggested by HA
In Arduino I have putted:
- #define MQTT_SERVER "ip address of HA"
- #define MQTT_PORT 1883
- #define MQTT_USERNAME "homeassistant"
- #define MQTT_PASSWORD "the one suggested by HA"

Then what I have to do in HA in order to read data from the inverter?
Thank you in advance!

Terafil

**Comment by double1968** [comment, marked as answer]
some examples:

`https://github.com/cmcgerty/Sofar2mqtt/discussions/3`

**Reply by terafilgit** [reply]
Thank you for your reply double1968, but the url is invalid

**Reply by double1968** [reply]
ok

**Comment by terafilgit** [comment]
No one knows how to use mqtt data in Home Assistant?

**Comment by cmcgerty** [comment]
Try this link again. There is a good example in the post.

https://github.com/cmcgerty/Sofar2mqtt/discussions/3#discussioncomment-1371254

Colin

On 13 Jun 2022, at 11:37, terafilgit ***@***.***> wrote:

﻿

No one knows how to use mqtt data in Home Assistant?

**Reply by terafilgit** [reply]
Thank you @double1968 and @cmcgerty, I finally integrate the inverter in HA

**Reply by Pierre8100** [reply]
Hello @terafilgit,
I think I have got the same inverter as you. Mine is a HYD6000-EP.
I have done the integration in HA and finaly found RS485 pin on inverter are the 5 and 6 (I am right ?)
My concern is tehre is no values returned except the status, that is "standby", all other informations are "0".

Is there something to change in the software ? Registers ? How did you proceed to get the information from your HYD6000 ?

Thanks

**Comment by cliffdude** [comment]
Hi,

The "ep" looks like a new version of the hyd lineup. Mine is a "es" version
and is similar spec but less powerfull on the DC side.
So your problem could be that Sofar has implemented a new set of registers
on the rs485 bus or you have a bad connection. I would think the first case
is more likely as you said you are getting status information that means
the communication is working, it is just not "looking" at the right data.
Perhaps sniffing the bus somehow would provide more info, maybe Colin could
chime in here...





A quinta, 22/09/2022, 07:33, Pierre8100 ***@***.***> escreveu:

**Reply by Pierre8100** [reply]
Thank you for the quick answer. I really appreciate.
I have the same feeling as you. Registers may be differents.
I have never perform sniffing on RS485, and I do not know how to proceed. So if someone show me how to, it will be great.
I have written to SofarSolar asking if they can share the registers used... I am waiting an answer.

**Comment by cliffdude** [comment]
FYI

I have not tried this yet but you now have an integration in HA

https://github.com/StephanJoubert/home_assistant_solarman

**Reply by Pierre8100** [reply]
Thank you, looks good but not tested yet, My HA is quite strange and lot of integration are not working... I do not know why. I should restart setup from beginning. UNfortunately not time yet. And As I have build the PCB , I would have been very proud to make it works with my inverter :)

**Comment by cliffdude** [comment]
You should find everything that you need in the thread below linked by
Colin. If you need anything specific I can help you, just let me know.

Best regards
Richard

**Comment by cliffdude** [comment]
I also use Colin's board. I like biilding that stuff, I also modded V1 of
his code to get a webserver with basic info instead of the oled as my
inverter is outside and I never look at it.
I tried the github imtegration I sent you but it did not work, probably my
firmware on the datalogger is too old. They do have some files with the
register configs that might help you.

A terça, 11/10/2022, 21:04, Pierre8100 ***@***.***> escreveu:

**Comment by Pierre8100** [comment]
Yes, you are right there are some intersting information in configuration file [https://github.com/StephanJoubert/home_assistant_solarman/blob/main/custom_components/solarman/inverter_definitions/sofar_lsw3.yaml](url)
I will try with these registers values. Maybe not this week as time is missing.

If I am thinking well, I have to make replacement like :
PV1 power : 0x0252 changed with 0x000A.

Good to try

Thank you @cliffdude  for all help you are providing ;)

**Reply by Loudo42** [reply]
Hi,
I have also a HYD6000EP, but I have not yet finished the config. I have to plug it to the inverter.

For the inverter register, I have found it for the new registery:
[https://www.facebook.com/groups/2477195449252168/posts/2935145736790468/?comment_id=2936011160037259&reply_comment_id=2936201263351582]

And in this page they also link to this:
[https://www.facebook.com/groups/2477195449252168/permalink/2895353217436387/](url)

Maybe you could try and let me know.

A question about the subscribing to Sofar2mqtt.
Where do you subcribe? is it in the configuration.yaml.
What is the command?

In my configuration.yaml I have added:
mqtt:
    broker: 192.168.1.96
    port: 1883
Am I right? If not could please you help me.

Regards,
Ludovic

**Reply by Pierre8100** [reply]
Hi Ludovic,

Just tested and it looks like to works ! maybe it is not working for all registry, but at least for some of them.
here information received in home assistant using registry found with your link : 
![image](https://user-images.githubusercontent.com/73114263/200193885-53895653-207a-424d-9802-17d46efd85b5.png)

It loooks to work good using these registry for HYD 6000 EP : 
#define SOFAR_REG_RUNSTATE 0x0404
#define SOFAR_REG_GRIDV 0x0498
#define SOFAR_REG_GRIDA 0x0499
#define SOFAR_REG_GRIDFREQ 0x0484
#define SOFAR_REG_BATTW 0x0606
#define SOFAR_REG_BATTV 0x0604
#define SOFAR_REG_BATTA 0x0605
#define SOFAR_REG_BATTSOC 0x0608
#define SOFAR_REG_BATTTEMP 0x0607
#define SOFAR_REG_GRIDW 0x0504
#define SOFAR_REG_LOADW 0x050c
#define SOFAR_REG_PVW 0x05c4
#define SOFAR_REG_PVDAY 0x0684
#define SOFAR_REG_EXPDAY 0x0690
#define SOFAR_REG_IMPDAY 0x068c
#define SOFAR_REG_LOADDAY 0x0688
#define SOFAR_REG_BATTCYC 0x0698
#define SOFAR_REG_PVA 0x0585
#define SOFAR_REG_INTTEMP 0x0420
#define SOFAR_REG_HSTEMP 0x041a
#define SOFAR_REG_PV1 0x0586
#define SOFAR_REG_PV2 0x0589

So, trial seems to be success.

regarding Yaml configuration, I received information broker, port, user and password support in configuration.yaml will be deprecated and not supported anymore. I have removed them from configuration.yaml.
Just for information when these parameters where in my configuration.yaml file, broker Ip was 127.0.0.1 (loopback adress as home assistant is hosting the broker).
I am not an expert, but I suppose configuration parameter are now directly set up during mqtt broker installation.

In my configuration.yaml, now I only have the sensor definition like this : 
mqtt:
  sensor:
      - name: "Sofar PV Today Consumption"
        state_topic: "Sofar2mqtt/state"
        value_template: '{{ value_json.today_consumption | multiply(0.01) | float }}'
        device_class: energy
        unit_of_measurement: 'kWh'
        state_class: total_increasing
      - name: "Sofar Grid Voltage"
        state_topic: "Sofar2mqtt/state"
        value_template: '{{ value_json.grid_voltage / 10 }}'
        unit_of_measurement: 'V'
        device_class: voltage
        state_class: total
      - name: "Sofar PV Today Generation"
        state_topic: "Sofar2mqtt/state"
        value_template: '{{ value_json.today_generation | multiply(0.01) | float }}'
        device_class: energy
        unit_of_measurement: 'kWh'
        state_class: total_increasing
      - name: "Sofar Battery SOC"
        state_topic: "Sofar2mqtt/state"
        value_template: '{{ value_json.batterySOC }}'
        device_class: battery
        state_class: total
      - name: "Sofar Battery Temp"
        state_topic: "Sofar2mqtt/state"
        value_template: '{{ value_json.battery_temp }}'
        device_class: temperature
        unit_of_measurement: '°C'
      - name: "Sofar Battery Charging/Discharging"
        state_topic: "Sofar2mqtt/state"
        value_template: >-
          {% if value_json.battery_power > (65535 / 2) %}
          {{ -(65535 - value_json.battery_power) * 10 }}
          {% else %}
          {{ (value_json.battery_power) * 10 }}
          {% endif %}
        unit_of_measurement: 'W'
        device_class: power
      - name: "Sofar State"
        state_topic: "Sofar2mqtt/state"
        value_template: >-
          {% if value_json.running_state == 0 %}
          {{ "Standby" }}
          {% elif value_json.running_state == 1 %}
          {{ "Check charge" }}
          {% elif value_json.running_state == 2 %}
          {{ "Charging" }}
          {% elif value_json.running_state == 3 %}
          {{ "Check discharge" }}
          {% elif value_json.running_state == 4 %}
          {{ "Discharging" }}
          {% elif value_json.running_state == 5 %}
          {{ "EPSState" }}
          {% elif value_json.running_state == 6 %}
          {{ "Fault" }}
          {% elif value_json.running_state == 7 %}
          {{ "Perminant Fault" }}
          {% else %}
          {{ "Undeterminated" }}
          {% endif %}
        unit_of_measurement: ''

**Reply by Loudo42** [reply]
Hi Pierre,
Glad to see it seems working. Let's wait rising sun to check if other sensor work.
I have some time to loose in different flights this week, so let me know the ones which do not work, I could check it in the modbus files of sofar if it can help.
And thank you for your advice.
I will try to plug it next week end.

**Reply by Pierre8100** [reply]
I have updated the configuration.yaml with MartinJSa proposal.
Some of information do not match.
Not a big deal but registers for grid voltage, grid current and few others are not giving feedback
information for inverter battery voltage and inverter battery current, needs to be corrected as coma is not a the right position.

Anyway, it is a good improvement and it works.

If anyone as the opportunity to help sniffing registry used, I think we are close to find a100% working system :)

**Reply by Loudo42** [reply]
Hi,
for the grid voltage and current, the value 0498 and 0499 are for phase S. My logger in solarman show a R before the value of grid (see attached picture) so if it is also your case maybe you could try:
#define SOFAR_REG_GRIDV 0x048D
#define SOFAR_REG_GRIDA 0x048E

If it is for phase T it should be
#define SOFAR_REG_GRIDV 0x04A3
#define SOFAR_REG_GRIDA 0x04A4

What are the other sensors which do not give datas. I could try to check it.

![Screenshot_20221109_000552_com igen xiaomaizhidian](https://user-images.githubusercontent.com/80635546/200675716-e314bb7d-3860-46e4-948d-58dc2c904ade.jpg)

**Reply by Pierre8100** [reply]
Thank you for your feedback.
Using the registers 0x048D et 0x048E is working.
I have the same configuration as you and the R values.

I think there is something wrong as these values are not grid values but home usage values.
The grid power register 0x0504 is not giving feedback, allways 0. 

House consumption power (0x050c) is also always 0
same for PV power (0x05c4) it stay at 0

I am interested also to get the voltage of PV1 and 2. I have not the registers for these values.

Thank you

**Reply by Loudo42** [reply]
Hi Pierre,

You could try these different value, I am not sure about it.
In my side I have reach to plug and get data's from the inverter. Next week end I will try to find the good registers by testing too. 
![Screenshot_20221114_081054](https://user-images.githubusercontent.com/80635546/201597451-5a8ef99a-7939-4c77-926c-7cb2163b7c9b.jpg)

**Comment by Loudo42** [comment]
Hi,

I succeed to find all the register for our sensor.
I did not find the Total PV power, but I have created in HA a sensor which is the sum of PV1 and PV2.

I have also added the total energy, but I was not able to publish all the sensors, I do not know why it is does not work. Maybe too much data to send.
So I have removed the Battery amps, the grid frequency and the daily battery charge and discharge.

If somebody know how publishing all thing, I will be gratefull.

The only issue is that the status stay all the time at 2-charging, even during discharging. So maybe the register is not good.
Is it working for you.

So I will have to create my rules.
For that do you know the exact code to send the command to the inverter?

Here is the list for my HYD 6000 EP:
#define SOFAR_REG_RUNSTATE 0x0404
#define SOFAR_REG_GRIDV 0x048D
#define SOFAR_REG_GRIDA 0x048E
#define SOFAR_REG_BATTW 0x0606
#define SOFAR_REG_BATTV 0x0604
#define SOFAR_REG_BATTA 0x0605
#define SOFAR_REG_BATTSOC 0x0608
#define SOFAR_REG_BATTTEMP 0x0607
#define SOFAR_REG_GRIDW 0x0488
#define SOFAR_REG_LOADW 0x04af
#define SOFAR_REG_PVDAY 0x0685
#define SOFAR_REG_EXPDAY 0x0691
#define SOFAR_REG_IMPDAY 0x068d
#define SOFAR_REG_LOADDAY 0x0689
#define SOFAR_REG_INTTEMP 0x0418
#define SOFAR_REG_HSTEMP 0x041a
#define SOFAR_REG_PV1 0x0586
#define SOFAR_REG_PV2 0x0589
#define SOFAR_REG_BATTCH 0x0697
#define SOFAR_REG_BATTDICH 0x069b
#define SOFAR_REG_PURCTOT 0x068f
#define SOFAR_REG_EXPTOT 0x0693
#define SOFAR_REG_LOADTOT 0x068b
#define SOFAR_REG_PVTOT 0x0687
//#define SOFAR_REG_PVW 0x05c4 DO NOT WORK SHOULD BE PV POWER TOT
//#define SOFAR_REG_GRIDFREQ 0x0484
//#define SOFAR_REG_BATDCH 0x0695 TODAY BAT CHARGE
//#define SOFAR_REG_BATDDICH 0x0699 TODAY BAT DISCHARGE
//0485  AC POWER GENERATED BY THE INVERTER
//0584  Voltage_PV1
//0587  Voltage_PV2

The list for the Json:
{
{ SOFAR_REG_RUNSTATE, "running_state" },
{ SOFAR_REG_GRIDV, "grid_voltage" },
{ SOFAR_REG_GRIDA, "grid_current" },
{ SOFAR_REG_BATTW, "battery_power" },
{ SOFAR_REG_BATTV, "battery_voltage" },
{ SOFAR_REG_BATTSOC, "batterySOC" },
{ SOFAR_REG_BATTTEMP, "battery_temp" },
{ SOFAR_REG_GRIDW, "grid_power" },
{ SOFAR_REG_LOADW, "consumption" },
{ SOFAR_REG_INTTEMP, "inverter_temp" },
{ SOFAR_REG_HSTEMP, "inverter_HStemp" },
{ SOFAR_REG_PV1, "Solarpv1" },
{ SOFAR_REG_PV2, "Solarpv2" },
{ SOFAR_REG_PVDAY, "today_generation" },
{ SOFAR_REG_EXPDAY, "today_exported" },
{ SOFAR_REG_IMPDAY, "today_purchase" },
{ SOFAR_REG_LOADDAY, "today_consumption" },
{ SOFAR_REG_BATTCH, "total_bat_charge" },
{ SOFAR_REG_BATTDICH, "total_bat_discharge" },
{ SOFAR_REG_PURCTOT, "total_purchase" },
{ SOFAR_REG_EXPTOT, "total_exported" },
{ SOFAR_REG_LOADTOT, "total_consumption" },
{ SOFAR_REG_PVTOT, "total_generation" },
//{ SOFAR_REG_BATTA, "battery_current" },
//{ SOFAR_REG_GRIDFREQ, "grid_freq" },
//{ SOFAR_REG_BATDCH, "today_bat_charge" },
//{ SOFAR_REG_BATDDICH, "today_bat_discharge" },
//{ SOFAR_REG_PVW, "solarPV" },
};

And the modification for the sensor in the HA's configuration.yaml :

  
    - name: "solar_power"  
      state_topic: "Sofar2mqtt/state"
      unit_of_measurement: "W"
      device_class: power
      state_class: measurement
      value_template: "{{ value_json.Solarpv1 *10+value_json.Solarpv2 *10 }}"
      #value_template: "{{ value_json.solarPV *10 }}"
      
    - name: "solar_power1"  
      state_topic: "Sofar2mqtt/state"
      unit_of_measurement: "W"
      device_class: power
      state_class: measurement
      value_template: "{{ value_json.Solarpv1 *10 }}"
      
    - name: "solar_power2"  
      state_topic: "Sofar2mqtt/state"
      unit_of_measurement: "W"
      device_class: power
      state_class: measurement
      value_template: "{{ value_json.Solarpv2 *10 }}"
     
    - name: "energy_total_batt_charge"  
      state_topic: "Sofar2mqtt/state"
      unit_of_measurement: "MWh"
      device_class: energy
      state_class: total_increasing
      value_template: "{{ value_json.total_bat_charge /10000 }}"
      # total_b
    
    - name: "energy_total_batt_discharge"  
      state_topic: "Sofar2mqtt/state"
      unit_of_measurement: "MWh"
      device_class: energy
      state_class: total_increasing
      value_template: "{{ value_json.total_bat_discharge /10000 }}"
      
    - name: "energy_total_imported"  
      state_topic: "Sofar2mqtt/state"
      unit_of_measurement: "MWh"
      device_class: energy
      state_class: total_increasing
      value_template: "{{ value_json.total_purchase /10000 }}"
      
    - name: "energy_total_exported"  
      state_topic: "Sofar2mqtt/state"
      unit_of_measurement: "MWh"
      device_class: energy
      state_class: total_increasing
      value_template: "{{ value_json.total_exported /10000 }}"    
      
    - name: "energy_total_consumption"  
      state_topic: "Sofar2mqtt/state"
      unit_of_measurement: "MWh"
      device_class: energy
      state_class: total_increasing
      value_template: "{{ value_json.total_consumption /10000 }}" 
            
    - name: "energy_total_generation"  
      state_topic: "Sofar2mqtt/state"
      unit_of_measurement: "MWh"
      device_class: energy
      state_class: total_increasing
      value_template: "{{ value_json.total_generation /10000 }}"

**Reply by Pierre8100** [reply]
Thank you very much @Loudo42, I am too busy curently to perform test and update, but I really apreciate you found register and give all teh details.
i am not an expert on github, but I think there is possibilities to push an updated file to integrate the specyfication of HYD inverter.

Anyway, thank you very much for the job done 👍
